File: 20241022/jueun/퍼즐게임챌린지.java
class Solution {
    int MAX = 100_001;
    
    int[] diffs;
    int[] times;
    long limit;
    
    public int solution(int[] diffs, int[] times, long limit) {
        this.diffs = diffs;
        this.times = times;
        this.limit = limit;
        
        return binarySearch(0, MAX);
    }
    
    private int binarySearch(int low, int high) {
        while(low+1 < high) {
            int mid = (low+high)/2;
            
            if(check(mid)) {
                low = mid;
            } else {
                high = mid;
            }
        }
        
        return high;
    }
    
    private boolean check(int level) {
        long time = 0;
        
        for(int i=0; i<diffs.length; i++) {
            long tryTimes = diffs[i] - level;
            
            if(tryTimes > 0) {
                time += (times[i-1]+times[i]) * tryTimes;
            }
            time += times[i];
        }
        
        //level 0에서 true -> time이 큼.
        //level MAX에서 false -> time이 작음..
        //true가 리턴되는 범위가 작게 (TTT FFF)
        return time > limit;
    }
}

File: 20241022/heemin/퍼즐_게임_챌린지.cpp
#include <string>
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

bool solution(int level, vector<int> diffs, vector<int> times, long long limit){
    
    long long total=0;

    for(int i=0;i<diffs.size();i++){
        int diff=diffs[i];
        int time=times[i];
        
        if( diff <= level ){
            total += time;
        }
        else{
            int cnt = diff-level;
            total += (time+times[i-1])*cnt;
            total += time;
        }
        
        if(total > limit){
            return false;
        }
    }
    return true;
}

int solution(vector<int> diffs, vector<int> times, long long limit) {
    int answer = 0;
    
    int start=1;
    int end = 100000;
    int mid=0;
    
    while(start <= end ){
        mid=(start+end)/2;
        
        if(solution(mid,diffs,times,limit)){
            answer=mid;
            end=mid-1;
        }
        else{
            start=mid+1;
        }
    }
    
    return answer;
}